Add length validation on name and description fields in wizards

mturley opened the pull request:
Closes #186.

Not adding any validation to the CSV step in this PR, that will be another PR and possibly another issue.

http://recordit.co/Vwt0VA2WQw

![vwt0va2wqw](https://user-images.githubusercontent.com/811963/39451704-8910c9be-4c9d-11e8-8102-d2be73085d4f.gif)

serenamarie125 commented on app/javascript/react/screens/App/Overview/screens/MappingWizard/__tests__/__snapshots__/index.test.js.snap:
```diff
@@ -23,7 +23,7 @@ Object {
     },
     "syncErrors": Object {
-      "name": "Required",
+      "name": "Required. Maximum is 24 characters.",
```
could be a nit and covered in a subsequent PR.  @vconzola I was just poking around and i wonder if we should state "Maximum length is 24 characters." what do ya think?

mturley replied:
I think either one works, "the maximum" for a string is obviously the max length, if we needed to save space. But we have plenty of horizontal space here so I agree that maybe "Maximum length is..." is better here.

vconzola commented:
@serenamarie Post-Summit we should do a thorough text review and include it there.

AllenBW commented on app/javascript/common/constants.js:
```diff
@@ -0,0 +1,8 @@
+export const MAX_LENGTH_NAMES = 24;
+export const MAX_LENGTH_DESCRIPTIONS = 128;
+
+export const nameHelp = `Maximum is ${MAX_LENGTH_NAMES} characters.`;
```
Might need to annotate these to support translation efforts 🤔 

Checkout the use of `sprintf`

AllenBW replied:
ps: love LOVE that they're all located in one place 😍

mturley replied:
I'll look into that now, I was a little worried about that actually. It wouldn't be the end of the world to hard-code the numbers into the strings. I'll look at `sprintf` first.

AllenBW commented on app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardGeneralStep/MappingWizardGeneralStep.js:
```diff
@@ -1,8 +1,17 @@
+  nameRequired,
+  nameTooLong,
+  descriptionTooLong
+} from '../../../../../../../../common/constants'; // Oh my
```
funny funny.. sooooooo these are all related... used in this one and only file... what if there was an optional bundled export (instead of each an every one cuz that could get insane long run)

mturley replied:
That's true. Haha. Started this pattern when there were only a few and it kinda got away from me. I'll clean that up.

AllenBW commented on app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardGeneralStep/MappingWizardGeneralStep.js:
```diff
@@ -11,14 +20,30 @@ const MappingWizardGeneralStep = props => (
+      help={__(nameHelp)}
+      validate={[
+        required({
+          msg: __(nameRequired)
```
OH you translate here... maybe earlier comment is hogwash... but then again... is one location is better than the other 🤔

mturley replied:
ohh, `__()` is the `sprintf` you meant? I think I did it here only to prevent a line break in the constants file. I can move it back if that's more clear. I'll do that, actually.

mturley replied:
@AllenBW was this `__()` sufficient to address your hogwash? or should I still not be using string template literals to put these numbers in these strings?

AllenBW replied:
Thinking its better to do the translation in the constants file with sprintf yeah, its the same as `__()` but allows for addition of variables...

mturley replied:
Ok i'll look at sprintf. The commits I just pushed fix everything but that.

AllenBW commented on app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardGeneralStep/PlanWizardGeneralStep.js:
```diff
@@ -1,10 +1,19 @@
+  nameRequired,
+  nameTooLong,
+  descriptionTooLong
+} from '../../../../../../../../common/constants'; // Oh my
```
hey this looks familiar.. looks a little WET... if only it could be DRY'd....

mturley replied:
yeah, yeah :) i'll go get the hair dryer, good old Cmd+Shift+F

mturley replied:
Wait, DRY is an acronym but did someone actually come up with a WET backronym?  Weakly Extracted Text?

AllenBW replied:
😆 https://en.wikipedia.org/wiki/Don't_repeat_yourself is one of the good ones

mturley replied:
haha yes, I know. haha. I prefer this wiki for coding mantras though. 😏 http://wiki.c2.com/?DontRepeatYourself

mturley commented:
Good point @michaelkro. Do you guys think we should do this (have max length be a component level thing and not just a validation thing) in this PR, or a separate one? Also, if it's at the component level, we could just stop them from typing more than the limit at all (and those over limit messages need not be in there at all). Open to suggestions there.

@serenamarie125 sorry I didn't share the recording before, thought I had. Just a moment.

mturley commented:
Recording of this PR in action: http://recordit.co/aEd8yjVuYo

I need to figure out how to embed these recordit videos directly in GitHub (they are real videos not gifs... There's a button to convert to gif and I could use that.. But that makes me sad from a file format standpoint :laughing:

AllenBW commented:
You honor me kind sir 🙇 💘 🍍

mturley commented:
Still looking at one minor code thing here, but I rebased and refactored this to use a component-level maxLength as suggested by @michaelkro.  Here's the updated recording of how it looks:

http://recordit.co/Vwt0VA2WQw

![vwt0va2wqw](https://user-images.githubusercontent.com/811963/39451694-80ac5b08-4c9d-11e8-86cd-999715b2991c.gif)

mturley commented:
Ok, I believe this PR is now fully ready to go. @priley86 in particular can you make sure what I'm doing with `onChange` in `FormField` is ok? I only call the injected `input.onChange` if there is a new value that is within the maximum. I considered calling it again with the current value instead of not calling it, but it would have no effect.

priley86 commented on app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardGeneralStep/MappingWizardGeneralStep.js:
```diff
@@ -3,6 +3,7 @@ import { Field } from 'redux-form';
 import { required } from 'redux-form-validators';
 import { Form } from 'patternfly-react';
 import { FormField } from '../../../../../common/forms/FormField';
+import { validation } from '../../../../../../../../common/constants'; // Oh my
```
these can be eliminated by using "shared module resolution". I missed that initially... feel free to file a "code improvement" issue and we can eliminate those in the future...
https://gist.github.com/ryanflorence/daafb1e3cb8ad740b346#shared-module-resolution

mturley replied:
Good point. Since this `../../../../` import pattern is already spread around this repo, let's not block this PR on it. Code improvement issue sounds great, i'll open one.

mturley replied:
By the way @priley86 I did open an issue for this: https://github.com/priley86/miq_v2v_ui_plugin/issues/270

AparnaKarve commented on app/javascript/common/constants.js:
```diff
@@ -0,0 +1,20 @@
+
+export const validation = {
+  name: {
+    help: sprintf(__(maxLengthIs), MAX_LENGTH_NAMES),
```
@mturley The i18n here will not work because the `gettext` parser looks for a `__('string')` format during string extraction.

Can you move the `__(` to the const above?

```js
const maxLengthIs = __('Maximum length is %s characters.');
```

mturley commented:
Thanks @AparnaKarve and @michaelkro, that issue is now resolved!

mturley commented:
Oh, sorry-- hang on. I misunderstood Aparna and moved the string literals down, duplicating them. I see what she meant, one more commit.

mturley commented:
There, @AparnaKarve now it's fixed the way you suggested.

AparnaKarve commented:
@mturley Thanks for fixing the i18n part - that looks good!

priley86 commented:
thanks @mturley !!! 👏 👏
